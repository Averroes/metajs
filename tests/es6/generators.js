var gen = function* () {
  try {
    for (var i = 0; i < 5; i++) {
      console.log("Iterator received", yield i);
    }
  }
  catch (e) {
    console.log("Iterator caught", e);
  }
  finally {
    console.log("Shutting down iterator");
  }
}

console.log("Generator 1");

var it = gen();
it.send();

while (true) {
  try {
    console.log(it.send("hi"));
  }
  catch (e) {
    if (e instanceof StopIteration) {
      console.log("iterator exhausted");
      break;
    }
    else
      throw e;
  }
}

console.log("Generator 2");

it = gen();
it.next();
console.log(it.close());

try {
  console.log(it.next());
}
catch (e) {
  if (e instanceof StopIteration) {
    console.log("iterator exhausted");
  }
  else
    throw e;
}

console.log("Generator 3");

it = gen();
it.next();
try {
  it.throw(new Error("injected error"));
}
catch (e) {
  if (e instanceof StopIteration)
    console.log("iterator exhausted");
  else
    throw e;
}
