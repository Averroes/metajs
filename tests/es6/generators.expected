Generator 1
Iterator received hi
1
Iterator received hi
2
Iterator received hi
3
Iterator received hi
4
Iterator received hi
Shutting down iterator
iterator exhausted
Generator 2
Shutting down iterator
undefined
iterator exhausted
Generator 3
Iterator caught [Error: injected error]
Shutting down iterator
iterator exhausted
